src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -179,68 +177,45 @@ protected virtual string GenerateUniqueId()
-                    chunk.Attributes,
-                    tagHelperDescriptor.TypeName,
-                    tagHelperVariableName,
-                    tagHelperDescriptor.Attributes,
-                    htmlAttributeValues);
             }
         }
 
-        private void RenderBoundHTMLAttributes(
-            IList<KeyValuePair<string, Chunk>> chunkAttributes,
-            string tagHelperTypeName,
-            string tagHelperVariableName,
-            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
-            Dictionary<string, string> htmlAttributeValues)
+        private void RenderTagHelperPropertyVerification(
+            IEnumerable<KeyValuePair<string, Chunk>> chunkAttributes,
+            IEnumerable<TagHelperDescriptor> tagHelperDescriptors)
         {
+            if (_designTimeMode)
+            {
+                // No need to render any dictionary verification in design time mode.
+                return;
+            }
+
             // Track dictionary properties we have confirmed are non-null.
             var confirmedDictionaries = new HashSet<string>(StringComparer.Ordinal);
 
-            // First attribute wins, even if there are duplicates.
-            var distinctAttributes = chunkAttributes.Distinct(KeyValuePairKeyComparer.Default);
-
-            // Go through the HTML attributes in source order, assigning to properties or indexers as we go.
-            foreach (var attributeKeyValuePair in distinctAttributes)
+            foreach (var tagHelperDescriptor in tagHelperDescriptors)
             {
-                var attributeName = attributeKeyValuePair.Key;
-                var attributeValueChunk = attributeKeyValuePair.Value;
-                if (attributeValueChunk == null)
+                foreach (var attributeDescriptor in tagHelperDescriptor.Attributes)
                 {
-                    // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
-                    // logged an error if it was a bound attribute; so we can skip.
-                    continue;
-                }
+                    if (attributeDescriptor.IsIndexer && confirmedDictionaries.Add(attributeDescriptor.PropertyName))
+                    {
+                        var attributeName = chunkAttributes.FirstOrDefault(
+                            chunkAttribute => attributeDescriptor.IsNameMatch(chunkAttribute.Key)).Key;
 
-                // Find the matching TagHelperAttributeDescriptor.
-                var attributeDescriptor = attributeDescriptors.FirstOrDefault(
-                    descriptor => descriptor.IsNameMatch(attributeName));
-                if (attributeDescriptor == null)
-                {
-                    // Attribute is not bound to a property or indexer in this tag helper.
-                    continue;
-                }
+                        // attributeName will be null in the case that the attribute is not utilized.
```

Comment:
The check is different. We used to iterate through the `chunkAttribute` and check to see if there's a descriptor associated. Now we're iterating the `TagHelperAttributeDescriptor`s and seeing if there's a `chunkAttribute` associated.